src/sharing/ShareWrapper.js

```diff
@@ -157,13 +154,37 @@ class ShareWrapperInner extends React.PureComponent<Props, State> {
       return;
     }
 
-    this.setSending();
+    this.setState({ sending: true });
     showToast(_('Sending message…'));
     if (sharedData.type === 'file') {
       const { files } = this.state;
       for (let i = 0; i < files.length; i++) {
-        const response = await api.uploadFile(auth, files[i].url, files[i].name);
-        messageToSend += `\n[${files[i].name}](${response.uri})\n`;
+        const { url, name: fileName } = files[i];
+        try {
+          const response = await api.uploadFile(auth, url, fileName);
+          messageToSend += `\n[${fileName}](${response.uri})\n`;
+        } catch (errorIllTyped) {
+          const error: mixed = errorIllTyped; // https://github.com/facebook/flow/issues/2470
+
+          if (!(error instanceof Error)) {
+            logging.error('ShareWrapper: Unexpected non-error thrown');
+          }
+
+          let msg = undefined;
+          if (
+            error instanceof ApiError
+            && error.code === 413 // 413 Payload Too Large:
+            //   https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/.2Fuser_uploads.3A.20.60413.20Payload.20Too.20Large.60/near/1332958
+          ) {
+            msg = _('This file is too large.');
+          } else if (error instanceof Error && error.message.length > 0) {
+            msg = error.message;
+          }
```

Comment:
That gives me a related thought: let's have this generic `Error` case, but before this let's say something like: if it's an `ApiError`, then `msg = _('The server said: {errorMessage}', { errorMessage: error.message })`.

That way again the attribution is clear.

Reply:
Makes sense. I put `\n\n` in this revision for consistency with a similar message we show in the login flow: `The server at {realm} said:\n\n{message}`